Reconnection logic

nytamin opened the pull request:
This PR adds several major features, as well as some minor fixes.

(This is a continuation of #10 )

## BREAKING: Yeelight.sendCommand now rejects promise if response from bulb is an error
Before, the sendCommand method always resolved the promise, no matter if the response from the bulb was an affirmative or an error. I've changed this so it rejects the promise if the bulb returns an error.

This is a breaking change, because it affects the interface to the consumer of the library.

## Infinite discovery
I wanted to be able to set the `Discover` class to continously sweep the network after devices. I reworked the logics in `Discover.start()` so that the promise won't resolve/reject but instead continue to sweep infinitely (or until `Discover.destroy()`
To do infinite discovery:
```
const discover = new Discover({
    limit: 0,
    timeout: 0
});
discover.on("deviceAdded", (device) => {
})
discover.start()
```
## Connection status & ping
I've added logic so that the Yeelight class emits "connected" & "disconnected" statuses.
Keeping track of the connection status by pinging the bulb regularly.

## Automatic reconnection
I've added logics for detecting when a connection has closed/failed and methods for reconnecting.
To activate:
```
const yeelight = new Yeelight({
    lightIp: '127.0.0.1'
});
yeelight.autoReconnect = true
yeelight.on('connected', () => {
    // This will change the light color every time the light is reconnected
    yeelight.setBright(1)
    .catch(e => console.log('error in setBright', e))
    yeelight.setRGB(new Color(
        Math.round(Math.random()* 255),
        Math.round(Math.random()* 255),
        Math.round(Math.random()* 255)
    ), 'sudden')
    .catch(e => console.log('error in setRGB', e))
})
yeelight.connect()
```

nytamin committed:
fix: handle multiple received messages

nytamin committed:
fix: wrong formula for calculating color (rgb = r * 65536 + g * 256 + b)

nytamin committed:
fix: bug in device discovery, allowing event 'deviceAdded' to be emitted multiple times

nytamin committed:
fix: discover event 'deviceAdded' could be emitted multiple times for the same device, if the fallback scanByIp was run. Added a check using host & port, to prevent this.

nytamin committed:
chore: fixed color test

nytamin committed:
feat: reworked discover class, to allow for infinite discovery

nytamin committed:
fix: add default duration to some methods, because the bulb behaved strange if it was missing

nytamin committed:
feat: add autoReconnect functionality to the Yeelight class

nytamin committed:
fix: use host & port instead of id to defermine if to emit new device
If a bulb is disconnected and then reconnected it might have gotten a new IP-adress. In that case the "deviceAdded" event should fire, so the consumer can establish a new connection

nytamin committed:
fix: port should be optional, use default port if omitted

nytamin committed:
feat: sendCommand now rejects promise if response from bulb is an error
THIS IS A BREAKING CHANGE

nytamin committed:
feat: Added ping method which pings the bulb regurarly, in order to determine if the connection has been disconnected

nytamin committed:
fix: reworked socket initialization due to possible event listener leak

nytamin committed:
feat: implemented config.filter function (already defined in interface)

nytamin committed:
chore: removed process.on("uncaughtException"), this should be handled by a consumer, not the library

nytamin committed:
chore: update typings to reflect actual return types

nytamin committed:
feat: added option to connect to a bulb by using the ID of the bulb only

nytamin commented:
More features:
## Connect to bulb using Id only
Added convenience function, so it's possible to connect to a bulb, using the ID of the bulb
Example:
```
const yeelight = new Yeelight({
    lightId: '0x0000000007c9505f' // whoa, by id only? amaaazing!
});
yeelight.autoReconnect = true
yeelight.on('connected', () => {
    // This will change the light color every time the light is reconnected
    yeelight.setRGB(new Color(
        Math.round(Math.random()* 255),
        Math.round(Math.random()* 255),
        Math.round(Math.random()* 255)
    ), 'sudden')
    .catch(e => console.log('error in setRGB', e))
})
yeelight.connect()
.catch(e => console.log('error in connect', e))
```

nytamin committed:
fix: lightId should be optional

nytamin committed:
chore: linting & cleanup

nytamin committed:
feat: add default value ("sudden") for transitions

nytamin committed:
fix: fixed setMusic host parameter

nytamin committed:
fix: ping should return null, not void

nytamin committed:
chore: added tests for all methods

nytamin committed:
chore: attempt to fix race-condition in unit-tests (causing socket reply data to bleed between tests)

nytamin committed:
chore: test-mock-socket server should not return null-data

nytamin committed:
chore: another try on fixing strange behaviour in travisCI tests

nytamin committed:
chore: add test for discover

samuraitruong commented:
Lets me know when you happy to merge this. Then we can release a new version. Make sure you update index.d.ts. Thanks

samuraitruong commented on src/models/color.ts:
```diff
@@ -11,6 +11,6 @@ export class Color {
     public getValue() {
         // tslint:disable-next-line:no-bitwise
-        return this.red * 65535 + this.green * 256 + this.blue;
+        return this.red * 0x10000 + this.green * 0x100 + this.blue;
```
I prefer not using bitwise, just correct the number above

nytamin replied:
fair enough, fixed! :)

nytamin committed:
chore: decimal representation of numbers

nytamin committed:
chore: add contributor

nytamin committed:
chore: removed index.d.ts and added declaration to ts-compiler instead

nytamin committed:
chore: add .npmignore file, so only build artifacts are included in the npm package

nytamin committed:
Merge branch 'develop' into feature/reconnection

# Conflicts:
#	src/discover.ts

nytamin commented:
The placement of index.d.ts looked a bit off, usually it's located together with the other generated js-files in /lib. I removed it and instead added an option in tsconfig.json, so it will be generated automatically when building.

I've also added an .npmignore, so only the build artifacts are published to npm (the source-files should not be included).

I'm happy with this now, feel free to merge it :)
